### PR TITLE
Match duplicate payments considering timezone differences and fix deletion of recurring payment duplicates

### PR DESCRIPTION
**Pull request by clarissalimab**
Problems found:
- OffsitePayments don't belong to donations with multiple payments, so not all payment dupes were deleted;
- OffsitePayments that already matched should not exist anymore, so it doesn't make sense that a `PaymentDupeStatus.matched = true` exists for them. They should've been deleted, but they weren't because of the bug mentioned on the first line. We fix that by running this:

```ruby
    PaymentDupeStatus.joins('LEFT JOIN payments ON payment_dupe_statuses.payment_id = payments.id').where("matched = true AND kind = 'OffsitePayment'").find_each do |ps|
        rd_match = PaymentDupeStatus.where('? = ANY(matched_with_offline)', ps.payment_id).first
        rd_match.matched = false
        rd_match.save!
        ps.matched = false
        ps.save!
    end
```

 - The OffsitePayments were created at midnight from the nonprofit's timezone. I suspect that when it was included on E-Tap, it was on UTC, so when they were imported back to CC, they ended up in the wrong date. Example:

| Activity | Date and time |
|:--------:|:-------------:|
| Payment created on CC | 2021-05-25 01:00:00 |
| Payment imported on E-Tap | 2020-05-24 20:00:00 (imported converted to nonprofit's timezone) |
| Payment imported back to CC | 2020-05-24 05:00:00 (which corresponds to midnight in the nonprofit's timezone) |

[Example in production](https://github.com/CommitChange/tix/issues/3615#issuecomment-1044963339).

### CODE REVIEW

File: lib/payment_dupes.rb [hunk 1]
```diff
@@ -134,24 +170,23 @@ def self.remove_payment_dupes(np_id, designations_to_become_comments)
...
                                         # we can do the copies once
-                                        copy_comment(offsite, online)
+                                        copy_comment(offsite, online, designations_to_become_comments)
                                         copy_dedication(offsite, online)
                                         copy_designation(offsite, online, designations_to_become_comments)
                                         deleted_payments.concat(temp_duplicate_payments)
                                         # deletes matching offsites here
-                                        p = Payment.find(temp_offsite_matches.first)
-                                        d = p.donation
-                                        d.payments.where(kind: 'OffsitePayment').each do |op|
```
Comment (clarissalimab) [inline lib/payment_dupes.rb]: This would not work because an offsite payment would not belong to a donation that had multiple other payments.

File: lib/payment_dupes.rb [hunk 2]
```diff
@@ -43,36 +47,65 @@ def self.can_copy_designation?(src, target, designations_to_become_comments)
...
                 # - the offsite donation has the same date as the online payment
                 # - there is a journal entry item for the offsite payment
-                
-                donation_or_ticket_payments = s.payments.not_matched.includes(:donation).where("(kind = 'Donation' OR kind = 'Ticket' OR kind = 'RecurringDonation') AND gross_amount = ? AND to_char(date, 'YYYY-MM-DD') = ?", offsite.gross_amount, offsite.date.strftime('%Y-%m-%d'))
+                donation_or_ticket_payments = s.payments.not_matched.includes(:donation).joins(
+                    'LEFT JOIN nonprofits ON payments.nonprofit_id = nonprofits.id'
+                ).where(
+                    "(kind = 'Donation' OR kind = 'Ticket' OR kind = 'RecurringDonation')
+                    AND (gross_amount = ? OR net_amount = ?) AND
+                    (to_char(timezone(COALESCE(nonprofits.timezone, \'UTC\'), timezone(\'UTC\', date)), 'YYYY-MM-DD') = ? OR to_char(date, 'YYYY-MM-DD') = ?)", offsite.gross_amount, offsite.gross_amount, offsite.date.strftime('%Y-%m-%d'), offsite.date.strftime('%Y-%m-%d'))
```
Comment (clarissalimab) [inline lib/payment_dupes.rb]: Transforms the online payment date into the nonprofit's timezone to match offsite payments that were exported in UTC and then imported using the nonprofit's timezone, but also matches both in UTC (as explained in the description).

File: lib/payment_dupes.rb [hunk 3]
```diff
@@ -43,36 +47,65 @@ def self.can_copy_designation?(src, target, designations_to_become_comments)
...
+                comment = target.donation.comment
+                designations_already_copied_to_comment.each do |d|
+                    comment = target.donation.comment.gsub(" \nDesignation: #{d}", "")
+                    comment = target.donation.comment.gsub("Designation: #{d}", "")
+                end
+                return true if (source.donation.comment.blank? || comment == source.donation.comment)
+            else
+                return false
+            end
+        end
```
Comment (clarissalimab) [inline lib/payment_dupes.rb]: Because we had some matches already that were not deleted, some comments would conflict because of the designations that were copied as comments. So we are comparing removing the designations that were copied.

File: lib/payment_dupes.rb [hunk 4]
```diff
@@ -43,36 +47,65 @@ def self.can_copy_designation?(src, target, designations_to_become_comments)
...
 
     def self.remove_payment_dupes(np_id, designations_to_become_comments)
         deleted_payments = []
         nonprofit = Nonprofit.find(np_id)
         etap_id_cf = CustomFieldMaster.find_by(name: 'E-Tapestry Id #').id
-        supp = nonprofit.supporters.not_deleted.joins(:custom_field_joins).where('custom_field_joins.custom_field_master_id = ?', etap_id_cf).references(:custom_field_joins)
-        
+        supp = nonprofit.supporters.not_deleted.joins(:custom_field_joins).where(
+            'custom_field_joins.custom_field_master_id = ?', etap_id_cf
+        ).references(:custom_field_joins)
```
Comment (clarissalimab) [inline lib/payment_dupes.rb]: Just formatting change to read better.

File: lib/payment_dupes.rb [hunk 5]
```diff
@@ -110,7 +143,10 @@ def self.remove_payment_dupes(np_id, designations_to_become_comments)
                                     temp_duplicate_payments = []
                                     temp_offsite_matches = []
                                     recurring_payments.find_each do |recurring_payment|
-                                        equivalent_offsite = s.payments.not_matched.where("kind = 'OffsitePayment' AND gross_amount = ? AND to_char(payments.date, 'YYYY-MM-DD') = ?", recurring_payment.gross_amount, recurring_payment.date.strftime('%Y-%m-%d')).joins(:journal_entries_to_item)
+                                        equivalent_offsite = s.payments.not_matched.where(
+                                            "kind = 'OffsitePayment' AND (gross_amount = ? OR gross_amount = ?) AND (to_char(payments.date, 'YYYY-MM-DD') = ? OR to_char(payments.date, 'YYYY-MM-DD') = ?)",
+                                            recurring_payment.gross_amount, recurring_payment.net_amount, recurring_payment.date.in_time_zone(nonprofit.timezone).strftime('%Y-%m-%d'), recurring_payment.date.strftime('%Y-%m-%d')
+                                        ).joins(:journal_entries_to_item)
```
Comment (clarissalimab) [inline lib/payment_dupes.rb]: Compares the offsite date with the online date in UTC and in the nonprofit's timezone.

File: lib/payment_dupes.rb [hunk 6]
```diff
@@ -43,36 +47,65 @@ def self.can_copy_designation?(src, target, designations_to_become_comments)
...
+            else
+                return false
+            end
+        end
         target.donation.comment = source.donation.comment
         target.donation.save!
     end
 
-    def self.can_copy_comment?(source, target)
+    def self.can_copy_comment?(source, target, designations_to_become_comments)
```
Comment (wwahammy) [inline lib/payment_dupes.rb]: Where is "designations_to_become_comments" coming from originally?

File: lib/payment_dupes.rb [hunk 7]
```diff
@@ -43,36 +47,65 @@ def self.can_copy_designation?(src, target, designations_to_become_comments)
...
         return true if source.donation.comment.blank?
         return true if target.donation.comment.present? && (source.donation.comment.blank? || target.donation.comment == source.donation.comment)
-        return false if target.donation.comment.present?
+        if target.donation.comment.present? 
+            if designations_to_become_comments.any? { |d| target.donation.comment.include?(d) }
+                designations_already_copied_to_comment = designations_to_become_comments.select { |d| target.donation.comment.include?(d) }
+                comment = target.donation.comment
+                designations_already_copied_to_comment.each do |d|
+                    comment = target.donation.comment.gsub(" \nDesignation: #{d}", "")
+                    comment = target.donation.comment.gsub("Designation: #{d}", "")
```
Comment (wwahammy) [inline lib/payment_dupes.rb]: Why are these both here?

File: lib/payment_dupes.rb (hunk 2 above)
Comment (wwahammy) [reply lib/payment_dupes.rb]: I'm not sure I understand why all this converting is going on. What are the actual conditions in which you want to match?

File: lib/payment_dupes.rb (hunk 5 above)
Comment (wwahammy) [reply lib/payment_dupes.rb]: I have similar questions as above, I don't understand what the conditions actually are.

File: lib/payment_dupes.rb (hunk 1 above)
Comment (wwahammy) [reply lib/payment_dupes.rb]: When you say it "would not work", what do you mean?

Comment (clarissalimab) [reply lib/payment_dupes.rb]: It means it didn't work to delete all the offsite payments because a donation doesn't have multiple offsite payments.

Comment (wwahammy) [reply lib/payment_dupes.rb]: As in, it doesn't actually do anything?

File: lib/payment_dupes.rb (hunk 2 above)
Comment (clarissalimab) [reply lib/payment_dupes.rb]: The examples in production I linked help to understand: https://github.com/CommitChange/tix/issues/3615#issuecomment-1044963339

And also the table that I did in the description to describe what's going on.

When the nonprofit makes an export, the dates and times show up in the nonprofit's timezone, and when the payments were imported back to us, we didn't consider as if the date was in the nonprofit's timezone and added the nonprofit's timezone on top of it. That caused many offline payments to have a different date than the corresponding online payment.

BUT, it was not always this behavior, I remember previously the export dates would always be in UTC and if they were importing our payments to E-Tapestry from exports generated before that change, then the date is correct.

To summarize, we want to match:
* Online payments dates in the nonprofit's timezone with offsite payments "in UTC": to match the offsite ones that were exported in the nonprofit's timezone and then were imported back as 00:00 on nonprofit's timezone (5:00 UTC for the nonprofit we're talking about).
* Online payments dates in UTC with offsite payments in actual UTC: to match the offsite ones that were exported in UTC and imported back as 00:00 on nonprofit's timezone (5:00 UTC for the nonprofit we're talking about).

File: lib/payment_dupes.rb (hunk 1 above)
Comment (clarissalimab) [reply lib/payment_dupes.rb]: It deleted 1 payment only.

File: lib/payment_dupes.rb (hunk 2 above)
Comment (wwahammy) [reply lib/payment_dupes.rb]: But are the conditions that they are either at UTC midnight or at the nonprofit's midnight?

File: lib/payment_dupes.rb (hunk 6 above)
Comment (clarissalimab) [reply lib/payment_dupes.rb]: https://github.com/CommitChange/houdini/pull/268/files#diff-0dbca3d4691fa54ac9678cda39b5ec2e9d18eed195ffa425fe2adbed140d82d5R88

File: lib/payment_dupes.rb (hunk 7 above)
Comment (clarissalimab) [reply lib/payment_dupes.rb]: Because when a donation doesn't have a comment, we just copy the designation without a \n, but if it has a comment already, we copy with a \n. Those two lines make sure we're removing both cases. See: https://github.com/CommitChange/houdini/pull/268/files#diff-0dbca3d4691fa54ac9678cda39b5ec2e9d18eed195ffa425fe2adbed140d82d5R23-R27

File: lib/payment_dupes.rb (hunk 5 above)
Comment (clarissalimab) [reply lib/payment_dupes.rb]: See this comment https://github.com/CommitChange/houdini/pull/268/files#r810294347

File: lib/payment_dupes.rb (hunk 2 above)
Comment (clarissalimab) [reply lib/payment_dupes.rb]: I didn't understand. They don't consider the time because the offsite payment time is always at the nonprofit's midnight, we don't compare the time, is that what you're asking?

Comment (wwahammy) [reply lib/payment_dupes.rb]: I guess I understand. I think the code could be clearer but if the tests pass, let's go with it.

File: lib/payment_dupes.rb (hunk 7 above)
Comment (wwahammy) [reply lib/payment_dupes.rb]: Alright.

File: lib/payment_dupes.rb (hunk 6 above)
Comment (wwahammy) [reply lib/payment_dupes.rb]: Sorry, I mean where do they come from originally?

Comment (clarissalimab) [reply lib/payment_dupes.rb]: Could you explain it in another way? `designations_to_become_comments` is a list of designations that we're moving to a comment because they are conflicting with the online payment designation but shouldn't be a problem, but because there was a bug in the payments deletion, the donations already had the designation as a comment.

Comment (wwahammy) [reply lib/payment_dupes.rb]: So the `designations_to_become_comments` are things we pass in manually?

Comment (clarissalimab) [reply lib/payment_dupes.rb]: Yes, take a look at step 8: https://github.com/CommitChange/tix/issues/3615#issuecomment-1034458827

**Review by wwahammy — APPROVED** [review]
If the tests pass and you're confident, I'm good with it.
